Introducing extensions for memcpy family of memory operations (#38)
ARM-software/acle: main/design_documents/__ARM_FEATURE_MOPS.rst

=============================================================

> Introducing extensions for memcpy family of memory operations

Reviewer: ```suggestion
Design Document for __ARM_FEATURE_MOPS
```

=============================================================

Following the announcement of the `2021 Architecture Extensions
<https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/arm-a-profile-architecture-developments-2021>`_,

> I'd like to propose these aditions to the ACLE to support the new instructions

Reviewer: Maybe avoid addressing this in first person, but use a generic "we"?

`We are proposing these additions to the ...`

> that standardise the memcpy family of operations.

Reviewer: Please mention `Armv8.8-A and Armv9.3-A` somewhere in this paragraph to make it clear which are the version of the architecture being mentioned here.

> Here's the rationale behind the contents:

Reviewer: Remove

Feature test macro
##################

> In order for users to determine the availability of the new memory operation
> instructions, as well of new potential intrinsics related to them, we are

Reviewer: ```suggestion
In order for users to determine the availability of the memory instructions introduced in Armv8.8-A and Armv9.3-A, as well of new potential intrinsics related to them, we are
```

proposing the addition of the following feature test macro, to be defined to 1
if the new extension is available:

::

  __ARM_FEATURE_MOPS

Intrinsics
##########

> Given that the purpose of the new instructions is not only optimizing but also

Reviewer: ```suggestion
Given that the purpose of the memory instructions in Armv8.8-A and Armv9.3-A is not only optimizing but also
```

standardising the software implementation of memcpy, memmove and memset
operations, the general use of these instructions should be covered by the code
generation of compilers and library implementations. Therefore, we feel there
might be no motivation to introduce new intrinsics covering this scenario.

> The new extension also introduces a few variations of such instructions covering

Reviewer: ```suggestion
Armv8.8-A and Armv9.3-A also introduce a few variations of such instructions covering
```

Also, what do you mean with `such instructions`? memcpy, memmove and memset? If so, these are better referred as functions, not instructions?

the following cases:

> * Privileged/unprivileged reads and writes
> * Temporal/non-temporal reads and writes
> * Tag setting variation of the memset operation

Reviewer: ```suggestion
* Privileged/unprivileged reads and writes.
* Temporal/non-temporal reads and writes.
* Tag setting variation of the memset operation.
```

Privileged/unprivileged variations
==================================

At first sight, there doesn't seem to be enough motivation to grant new

> intrinsics to cover those variations. Regular usage should be covered by

Reviewer: ```suggestion
intrinsics to cover those variations. Regular usage is covered by
```
